tachyon/crypto/sumcheck/multilinear/data_structures.h

```diff
@@ -0,0 +1,173 @@
+#ifndef TACHYON_CRYPTO_SUMCHECK_MULTILINEAR_DATA_STRUCTURES_H_
+#define TACHYON_CRYPTO_SUMCHECK_MULTILINEAR_DATA_STRUCTURES_H_
+
+#include <algorithm>
+#include <memory>
+#include <numeric>
+#include <vector>
+
+#include "absl/container/node_hash_map.h"
+
+#include "tachyon/base/buffer/copyable.h"
+#include "tachyon/base/logging.h"
+#include "tachyon/math/polynomials/multivariate/multilinear_extension.h"
+
+namespace tachyon {
+
+namespace crypto::sumcheck {
+
+// This data structure  will be used as the verifier key.
+struct TACHYON_EXPORT PolynomialInfo {
+  size_t max_multiplicands;
+  size_t num_variables;
+
+  bool operator==(const PolynomialInfo& other) const {
+    return max_multiplicands == other.max_multiplicands &&
+           num_variables == other.num_variables;
+  }
+  bool operator!=(const PolynomialInfo& other) const {
+    return !operator==(other);
+  }
+  // TODO(ashjeong): For better unit testing
+  // PolynomialInfo Random() {}
+};
+
+template <typename Evaluations>
+class ListOfProductsOfPolynomials;
+
+template <typename Evaluations>
+class ListOfProductsOfPolynomials<math::MultilinearExtension<Evaluations>> {
+ public:
+  using F = typename math::MultilinearExtension<Evaluations>::Field;
+
+  struct Term {
+    F coefficient;
+    std::vector<size_t> indexes;
+  };
+
+  ListOfProductsOfPolynomials() = default;
+  explicit ListOfProductsOfPolynomials(size_t num_variables)
+      : num_variables_(num_variables) {}
+  ListOfProductsOfPolynomials(ListOfProductsOfPolynomials& other) = default;
+  ListOfProductsOfPolynomials& operator=(ListOfProductsOfPolynomials& other) =
+      default;
+
+  // Adds a Term to the list
+  void AddTerm(
+      F& coefficient,
+      std::vector<std::shared_ptr<math::MultilinearExtension<Evaluations>>>&
+          evaluations) {
+    CHECK(!evaluations.empty());
+    size_t num_evaluations = evaluations.size();
+    max_evaluations_ = std::max(max_evaluations_, num_evaluations);
+    std::vector<size_t> indexed_evaluations;
+    indexed_evaluations.reserve(num_evaluations);
+    for (std::shared_ptr<math::MultilinearExtension<Evaluations>> evaluation :
+         evaluations) {
+      CHECK_EQ(evaluation->Degree(), num_variables_);
+      const math::MultilinearExtension<Evaluations>* evaluation_ptr =
+          evaluation.get();
+      auto it = raw_pointers_lookup_table_.find(evaluation_ptr);
+      if (it != raw_pointers_lookup_table_.end()) {
```

Comment:
Please use `base::Contains`

```suggestion
      if (base::Contains(raw_pointers_lookup_table_, evaluation_ptr))
```